### PR TITLE
site-logo block: Use option instead of theme-mod

### LINKED ISSUE #32065: Site Logo Block: logo data source and block UX
The Site Logo block is planned for release in WordPress 5.8.

It was created to use a dedicated option (`site_logo`) that interoperated with the `custom-logo` `theme_mod` used in the Customizer:

- If a `site_logo` option is set from the block, that logo is shown in the Customizer in place of the `theme_mod` (by hooking into `theme_mod_custom_logo`)
- If the `custom-logo` theme mod is updated or deleted in the Customizer, that logo is copied into the `site_logo` option or the `site_logo` option is deleted

[Related code](https://github.com/WordPress/gutenberg/blob/f3a576314c647465ef5ffac8eb6cf292cda91059/packages/block-library/src/site-logo/index.php#L67)

---
In preparation for merging Gutenberg into Core for the 5.8 release, [the block was revised](https://github.com/WordPress/gutenberg/pull/31994) to use the `custom-logo` `theme_mod` directly, rather than a separate option. (I believe this was prompted by a failing Core unit test, but I don't know the full context.)

This has a negative impact on the UX of the Site Logo block, in that the logo now disappears from the block when switching themes (because the block's logo setting is now theme specific).

https://user-images.githubusercontent.com/1699996/119033809-48e0c600-b973-11eb-8cc6-cb87b1b28c76.mov

While this is considered normal for the Custom Logo option in the Customizer, I'm not convinced it's an acceptable experience for the Site Logo block. Having the logo setting disappear when switching themes seems broken to me. I'm hoping we can improve the UX before the block is released in Core.

I'm creating this issue to bring awareness of the change, it's impact, and discuss what the best path forward for the Site Logo block in WP 5.8.

How should the site logo block setting be stored?

What's the best way to balance backward compatibility with the `custom-logo` `theme_mod` and create good experience of using the block (including when switching themes)?

What can we achieve before the 5.8 release?

### LINKED ISSUE #32171: Site Logo Block doesn't load after previously uploading
## Description

With Gutenberg Version 10.7.0-rc.1, the Site Logo block now just shows as infinitely loading without an option to replace the image in the block toolbar. This is a regression from 10.6.2. 

## Step-by-step reproduction instructions

1. Use Gutenberg 10.6.2 and set a site logo in the Site Logo Block
2. Switch to Gutenberg Version 10.7.0-rc.1 and view where you previously added the Site Logo Block to see error. 

## Expected behaviour

Should be able to see the previous site logo that was uploaded and have the option to replace it. 

## Actual behaviour

The Site Logo Block doesn't show an option to replace it and continually shows as loading. 

## Screenshots or screen recording (optional)

https://user-images.githubusercontent.com/26996883/119401606-ebad8300-bc98-11eb-8e2c-4f4a8b8f198d.mov

## WordPress information
- WordPress version: 5.7.2
- Gutenberg version: Version 10.7.0-rc.1
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? TT1 Blocks

## Device information
- Device: Desktop
- Operating system: MacOS
- Browser: Chrome

### PR DESCRIPTION
**Pull request by aristath**
## Description

This PR reverts https://github.com/WordPress/gutenberg/pull/31994 and updates the theme-mod-to-option sync methods a bit.

## Types of changes
* Reverted https://github.com/WordPress/gutenberg/pull/31994
* Renamed the pre-existing sync functions and tweaked their inline docs a bit
* Added method to update the theme-mod when the option sets updated.
* Includes https://github.com/WordPress/gutenberg/pull/32187

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/site-logo/index.php [hunk 1]
```diff
@@ -72,4 +91,59 @@ function register_block_core_site_logo() {
...
+add_filter( 'theme_mod_custom_logo', '_override_custom_logo_theme_mod' );
+
+/**
+ * Updates the site_logo option when the custom_logo theme-mod gets updated.
+ *
+ * @param string $custom_logo The custom logo set by a theme.
+ *
+ * @return string The custom logo.
+ */
+function _sync_site_logo_to_custom_logo( $custom_logo ) {
```
Comment (creativecoder) [inline packages/block-library/src/site-logo/index.php]: I think this function name is reversed, as this is hooked into `pre_set_theme_mod_custom_logo`... should it be `_sync_custom_logo_to_site_logo` instead?

(and vise versa for the function below, as well)

**Review by TimothyBJacobs — APPROVED** [review]
Code LGTM!

**Review by creativecoder — COMMENTED** [review]
Thanks for this PR!

I tested these scenarios

- Set site_logo in Site Logo block: ✅ option is synced to custom_logo theme mod
- Set custom_logo in Customizer: ✅ option is synced to site_logo option
- Delete site_logo with `delete_option`: ❌ custom logo_theme mod is not deleted (but it does work when using `update_option` with an empty value)
- Delete custom_logo from the Customizer: ✅ site_logo option is deleted
- Switch themes: ✅ logo continues to display in Site Logo block (and in Customizer)

Functionally this looks good. We may want to consider hooking into `delete_option_site_logo` as well ... but see below for a caveat.

---

In testing this with some logging, I did notice that simultaneously hooking into `pre_set_theme_mod_custom_logo` and `update_option_site_logo` runs some extra updates:

Update `site_logo` from the Site Logo block

- As expected, `update_option_site_logo` action is triggered and calls the hooked function that runs `set_theme_mod` to sync to the `custom_logo` theme mod
- This also causes the `pre_set_theme_mod_custom_logo` action to run and call `update_option` again
- But it stops there, because the option value is the same as what's already set

Update custom_logo from the Customizer

- As expected `pre_set_theme_mod_custom_logo` action is triggered calls the hooked function that runs `update_option` to sync to the `site_logo` option
- This also causes  the `update_option_site_logo` action to run and call `set_theme_mod` again
- Then  the `pre_set_theme_mod_custom_logo` action is triggered a second time and calls `update_option` again
- But it stops there, because the option value is the same as what's already set

**Comment by colorful-tones** [comment]
I just took this PR for a quick spin in TwentyTwentyOne. I created a New Page and added Site Logo and able to attach media. Working nicely from editorial flow perspective 👍

File: packages/block-library/src/site-logo/index.php (hunk 1 above)
Comment (aristath) [reply packages/block-library/src/site-logo/index.php]: fixed in https://github.com/WordPress/gutenberg/pull/32229/commits/389aed8ad57f8ac1f5a9d01379f1408432101335 :+1:

**Comment by aristath** [comment]
> In testing this with some logging, I did notice that simultaneously hooking into pre_set_theme_mod_custom_logo and update_option_site_logo runs some extra updates:

fixed :+1:

**Comment by cpapazoglou** [comment]
Hi there! Just a heads up (if you already haven't noticed it), `register_setting()` is also affecting saving in `/wp-admin/options-general.php`.
https://github.com/WordPress/gutenberg/blob/27afc4d43f02713fd40f4a35b6f6f8fabc175d0f/lib/rest-api.php#L190-L217

Cause, it updates `$new_allowed_options` with the provided option. `options-general.php` awaits all` $new_allowed_options` to be present in the form and thus we were getting an empty `stylesheet`.

Probably, using `register_setting()` in a `rest_api_init` hook **as you already do in this PR** https://github.com/WordPress/gutenberg/pull/32229/files#diff-92dc6950bcaf54290625664019ea23b023cb3f156de19586612c8d3e3cf52adcR81
is a good solution 👏!

Related issue https://github.com/Automattic/wp-calypso/issues/53447

**Comment by creativecoder** [comment]
@cpapazoglou Just to be clear: it looks like the changes merged from this PR fix the issue, is that right?

**Comment by cpapazoglou** [comment]
> @cpapazoglou Just to be clear: it looks like the changes merged from this PR fix the issue, is that right?

Sure they do!

File: packages/block-library/src/site-logo/index.php [hunk 2]
```diff
@@ -72,4 +91,63 @@ function register_block_core_site_logo() {
...
+	if ( empty( $value ) ) {
+		remove_theme_mod( 'custom_logo' );
+	} else {
+		remove_filter( 'pre_set_theme_mod_custom_logo', '_sync_custom_logo_to_site_logo' );
+		set_theme_mod( 'custom_logo', $value );
+		add_filter( 'pre_set_theme_mod_custom_logo', '_sync_custom_logo_to_site_logo' );
+	}
+}
+
+add_action( 'update_option_site_logo', '_sync_site_logo_to_custom_logo', 10, 2 );
```
Comment (pablinos) [inline packages/block-library/src/site-logo/index.php:153]: We were fixing an issue last night where this code sync'd values from a conflicting site option in Jetpack. 

One possible solution might be to check that the current theme supports the `custom_logo` theme mod, and if not, then we don't register this action.

/cc @creativecoder @kraftbj @fullofcaffeine

**Pull request merged**